File: rewrite-core/src/test/java/org/openrewrite/ChecksumTest.java
/*
 * Copyright 2025 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite;

import org.junit.jupiter.api.Test;
import org.openrewrite.ipc.http.HttpSender;

import java.io.ByteArrayInputStream;
import java.io.UncheckedIOException;
import java.net.URI;
import java.nio.charset.StandardCharsets;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ChecksumTest {

    @Test
    void fromHexWithValidInput() {
        Checksum checksum = Checksum.fromHex("SHA-256", "abcdef01");
        assertThat(checksum.getHexValue()).isEqualTo("abcdef01");
    }

    @Test
    void fromHexRejectsNonHexInput() {
        assertThatThrownBy(() -> Checksum.fromHex("SHA-256", "<?xml version"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("not valid hexadecimal");
    }

    @Test
    void fromUriRejectsNonSuccessfulResponse() {
        HttpSender httpSender = request -> new HttpSender.Response(
                403,
                new ByteArrayInputStream("<?xml version=\"1.0\"?>".getBytes(StandardCharsets.UTF_8)),
                () -> {
                });

        assertThatThrownBy(() -> Checksum.fromUri(httpSender, URI.create("https://example.com/checksum.sha256")))
                .isInstanceOf(UncheckedIOException.class)
                .hasMessageContaining("403");
    }

    @Test
    void fromUriTrimsWhitespace() {
        HttpSender httpSender = request -> new HttpSender.Response(
                200,
                new ByteArrayInputStream("abcdef01\n".getBytes(StandardCharsets.UTF_8)),
                () -> {
                });

        Checksum checksum = Checksum.fromUri(httpSender, URI.create("https://example.com/checksum.sha256"));
        assertThat(checksum.getHexValue()).isEqualTo("abcdef01");
    }
}

File: rewrite-core/src/test/kotlin/org/openrewrite/ChecksumTest.kt
/*
 * Copyright 2022 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.openrewrite.marker.Markers
import org.openrewrite.remote.Remote
import java.net.URI
import java.net.URL
import java.nio.file.Paths

class ChecksumTest {
    private val gradle742 = "575098db54a998ff1c6770b352c3b16766c09848bee7555dab09afc34e8cf590"

    @Test
    fun checksum() {
        val checksum = Checksum.fromHex(
            "SHA-256",
            URL("https://services.gradle.org/distributions/gradle-7.4.2-wrapper.jar.sha256").readText()
        )
        assertThat(checksum.hexValue).isEqualTo(gradle742)
    }

    @Test
    fun remoteChecksum() {
        val remote = Remote
            .builder(
                Tree.randomId(),
                Paths.get("gradle/wrapper/gradle-wrapper.jar"),
                Markers.EMPTY,
                URI.create("https://services.gradle.org/distributions/gradle-7.4.2-wrapper.jar")
            )
            .checksum(
                "SHA-256",
                URI.create("https://services.gradle.org/distributions/gradle-7.4.2-wrapper.jar.sha256")
            )
            .build()

        assertThat(remote.checksum?.hexValue).isEqualTo(gradle742)
    }
}
